python/tskit/metadata.py

```diff
@@ -454,12 +482,19 @@ def enforce_fixed_properties(obj):
                 return [enforce_fixed_properties(j) for j in obj]
             elif type(obj) == dict:
                 ret = {k: enforce_fixed_properties(v) for k, v in obj.items()}
-                if ret.get("type") == "object":
+                if "object" in ret.get("type", []):
                     if ret.get("additional_properties"):
                         raise ValueError(
                             "Struct codec does not support additional_properties"
                         )
-                    ret["required"] = list(ret.get("properties", {}).keys())
+                    # To prevent authors having to list required properties the default
+                    # is that all without a default are required.
+                    if "required" not in ret:
+                        ret["required"] = [
+                            prop
+                            for prop, sub_schema in ret.get("properties", {}).items()
+                            if "default" not in sub_schema
+                        ]
```

Comment:
Can check here if any properties that *don't* have a default are in required too.

Reply:
That's checked on line 174